### ISSUE TITLE
Class template partial specialization with concept-auto non-type parameter is not supported

### ISSUE THREAD

**Issue by Fedr** [llvm/llvm-project#77377]
In this program
```
#include <concepts>

template <auto>
struct A;

template <std::same_as<int> auto p>
struct A<p> {};

A<0> a;
```
`A<auto>` is a class template with non-type parameter, and `A<std::same_as<int> auto>` is its specialization for `int` parameters. The program is accepted by GCC and MSVC, but Clang complains:
```
error: class template partial specialization is not more specialized than the primary template [-Winvalid-partial-specialization]
```
Online demo: https://godbolt.org/z/ee4TcPY5K

**Comment by zyn0217** [comment]
We have to compare the template parameter equivalence before going through the constraint check, per [[temp.func.order]p6.2.2](https://eel.is/c++draft/temp.func.order#6.2.2):

> Otherwise, if the corresponding template-parameters of the template-parameter-lists are *not* equivalent [temp.over.link]... neither template is more specialized than the other.

And the definition for 'equivalent' is at [[temp.over.link]p6.3](https://eel.is/c++draft/temp.over.link#6.3):

> Two template-parameters are equivalent under the following conditions:
>
> if they declare non-type template parameters, they have equivalent types **ignoring** the use of type-constraints for placeholder types

And this is implemented by https://github.com/llvm/llvm-project/commit/bc62fb9e1779043ba4ffa75bb2133cc670d4105c.

> Interestingly, it does work with a requires clause
> https://godbolt.org/z/bYM43r4Yr

I think this could also explain why using a requires expression does work -- the parameters are considered _equivalent_ that way.

(I'm not 100% sure the standard is intended to reject such cases. CC @zygoloid for some opinions.)

**Comment by zygoloid** [comment]
> And the definition for 'equivalent' is at [[temp.over.link]p6.3](https://eel.is/c++draft/temp.over.link#6.3):
> 
> > Two template-parameters are equivalent under the following conditions:
> > if they declare non-type template parameters, they have equivalent types **ignoring** the use of type-constraints for placeholder types
> 
> And this is implemented by [bc62fb9](https://github.com/llvm/llvm-project/commit/bc62fb9e1779043ba4ffa75bb2133cc670d4105c).

In this case, the template parameters are equivalent, so we should continue to [[temp.func.order]/6.4](https://eel.is/c++draft/temp.func.order#6.4) which orders the function templates based on their constraints.
